libsweep/src/sweep.cc

```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
   uint8_t args[2] = {0};
   sweep::protocol::integral_to_ascii_bytes(code, args);
 
-  sweep::protocol::error_s protocolerror = nullptr;
+  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args);
 
-  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args, &protocolerror);
+  sweep::protocol::response_param_s response;
+  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to send sample rate command");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
-  }
+void sweep_device_reset(sweep_device_s device) {
+  SWEEP_ASSERT(device);
+  SWEEP_ASSERT(!device->is_scanning);
 
-  sweep::protocol::response_param_s response;
-  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response, &protocolerror);
+  sweep::protocol::write_command(device->serial, sweep::protocol::RESET_DEVICE);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to receive sample rate command response");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
```

Comment:
As it is translating an exception into and ABI error, which is defined in `sweep.h` and no other part of the code has to care about this I don't see why this should be moved there.